### PR TITLE
Darlin serializations

### PR DESCRIPTION
**Pull request by DanieleDiBenedetto**

### CODE REVIEW

**Review by phoinic — APPROVED** [review]

File: src/data_structures.rs [hunk 1]
```diff
@@ -66,7 +91,7 @@ pub trait PCPreparedCommitment<UNPREPARED: PCCommitment>: Clone {
 
 /// Defines the minimal interface of commitment randomness for any polynomial
 /// commitment scheme.
-pub trait PCRandomness: Clone {
+pub trait PCRandomness: Clone + Debug + Eq + PartialEq + CanonicalSerialize + CanonicalDeserialize {
```
Comment (UlrichHaboeck75) [inline src/data_structures.rs:94]: Why do we need implementations for ``Eq`` and ``ParitalEq`` here?

Comment (UlrichHaboeck75) [reply src/data_structures.rs:94]: One guess: for the serialize-deserialize tests?

Comment (DanieleDiBenedetto) [reply src/data_structures.rs:94]: You answered yourself :D

File: src/ipa_pc/data_structures.rs [hunk 2]
```diff
@@ -216,27 +278,83 @@ pub struct Proof<G: AffineCurve> {
...
-    fn size_in_bytes(&self) -> usize {
-        to_bytes![self].unwrap().len()
+impl<G: AffineCurve> SemanticallyValid for Proof<G> {
+    fn is_valid(&self) -> bool {
+        self.l_vec.is_valid() &&
+            self.r_vec.is_valid() &&
+            self.final_comm_key.is_valid() &&
+            self.c.is_valid() &&
+            if self.hiding_comm.is_some() { self.hiding_comm.as_ref().unwrap().is_valid() } else { true } &&
+            if self.rand.is_some() { self.rand.as_ref().unwrap().is_valid() } else { true }
```
Comment (UlrichHaboeck75) [inline src/ipa_pc/data_structures.rs:288]: We can do some further consistency checks here: is the length of l_vec that of r_vec (if not tested before),  and does the length correspond the log len of the committer key of the PC scheme.

File: src/ipa_pc/data_structures.rs [hunk 3]
```diff
@@ -263,29 +383,85 @@ pub struct BatchProof<G: AffineCurve> {
...
 }
 
-impl<G: AffineCurve> BatchPCProof for BatchProof<G> {
-    fn size_in_bytes(&self) -> usize {
-        to_bytes![self].unwrap().len()
+impl<G: AffineCurve> SemanticallyValid for BatchProof<G> {
+    fn is_valid(&self) -> bool {
+        self.proof.is_valid() &&
+            self.batch_commitment.is_valid() &&
+            self.batch_values.iter().all(|(_, v)| v.is_valid())
```
Comment (UlrichHaboeck75) [inline src/ipa_pc/data_structures.rs:390]: Here, we could check if the number of ``batch_values`` correspond to the (overall) number of commitments (including the ``batch_commitment``)

File: src/lib.rs [hunk 4]
```diff
@@ -647,6 +654,9 @@ pub mod tests {
                 &rands,
                 Some(rng),
             )?;
+
+            test_canonical_serialize_deserialize(true, &proof);
+
```
Comment (UlrichHaboeck75) [inline src/lib.rs:659]: Not related to the current changes:
Why is the test function called ``bad_degree_bound_test()``? 
And why doesn't it use segmentation (the sampled polynomials are only of supported_degree).

File: src/ipa_pc/data_structures.rs (hunk 2 above)
Comment (DanieleDiBenedetto) [reply src/ipa_pc/data_structures.rs:288]: I can do the first check you mentioned. Regarding the second one I can't, cause I don't have access to the committer key

File: src/ipa_pc/data_structures.rs (hunk 3 above)
Comment (DanieleDiBenedetto) [reply src/ipa_pc/data_structures.rs:390]: I don't have access to the commitments here

**Review by UlrichHaboeck75 — CHANGES_REQUESTED** [review]
See my comments (the ``is_valid()`` implementations).

**Review by UlrichHaboeck75 — APPROVED** [review]

**Review by albertog78 — APPROVED** [review]

**Pull request merged**